README.md
[![Build Status](https://img.shields.io/travis/fabbricadigitale/proauth.js.svg?style=flat-square)](https://travis-ci.org/fabbricadigitale/proauth.js)
[![Code Coverage](https://img.shields.io/codecov/c/github/fabbricadigitale/proauth.js.svg?style=flat-square)](https://codecov.io/gh/fabbricadigitale/proauth.js)

proauth.js
==========

> Automagically attach OAuth2's tokens to HTTP requests.

**proauth.js** is an OAuth 2.0 client with session management functionality and ability to attach the `Authorization` header to HTTP requests made by the browser.
Basically, when *proauth.js* is used alongside your app, it stores OAuth's tokens then handles any *managed URL*s
(ie. endpoints that require authentication, that you can specify within settings).

Any request to a *managed URL* (ie. an ajax to your authenticated API endpoint) will be intercepted in a seamless way.

For each HTTP request to a *managed URL*, once tokens are obtained and stored, *proauth.js* will:
- Attach the `Authorization` header containing the current access token
- Refresh the token when got a `401` response and retry to fetch the request with the new token

So, you don't have to care anymore about handling OAuth.

[![Build Status](https://saucelabs.com/browser-matrix/proauth.svg)](https://saucelabs.com/u/proauth)

## Install

To install this library, you can simply run the following command:

`npm install proauth.js`

Easy peasy!

## Usage

Just include the following snippet within `<head>`:
```html
<script>
  proauth = {
    // Put your settings here!
  }
</script>
<script src="lib/default.es2015.js"></script>
```

To start a new session you have just to make an API call to the configured oauth endpoint or you can use the following commodity:
```js
proauth.client.login(<username>, <password>)
```
> Return a promise fulfilled when the response is obtained

If authentication is successful, tokens are stored within the `localStorage` and used for subsequent requests to *managed URLs*.

To end the session you can use:
```js
proauth.client.clearSession()
```
or you can clear the `localStorage` by yourself.


### Configuration

You can configure *proauth.js* changing properties of the global `proauth` variable, before loading the library.

For example:
```js
proauth = {
  swSrc: "/service-worker.js",
  swOptions: {},
  oauthUrl: "/oauth",
  oauthClientId: "proauth",
  sessionStorage: "localStorage",
  namespace: "",
  managedUrls: [
      "/"
  ]
}
```
> User settings must be set proior to load the library


The properties you can customize are:
* `swSrc`: path to the service worker that will be registered when `default` package is loaded by the browser. Set it to a falsey value to avoid the auto-registration of service worker. (default: **"/service-worker.js"**)
* `swOptions`: object that will be passed as 2nd parameter to [ServiceWorkerContainer.register()](https://developer.mozilla.org/en-US/docs/Web/API/ServiceWorkerContainer/register). Note that this setting has effect only if `swSrc` is a valid service worker URL. (default: **{}**)
* `oauthUrl`: URL of the OAuth service that will be used to negotiate tokens. (default: **"/oauth"**)
* `authClientId`: OAuth client that will be used during negotiation of tokens. (default: **"proauth"**)
* `sessionStorage`: where to store tokens and session informations. At the moment, only `localStorage` is supported. (default: **"localStorage"**)
* `namespace`: namespace that will be used to distinguish between sessions. (default: **document.origin && document.origin !== "null" ? document.origin : ""**)
* `managedUrls`: array of base URLs where to attach the *Authorization* header. (default: **["/"]**)


## How it works

This library implements a Client-Service model (both on browser side), also defining its own communication protocol.
The Client will communicate with the Service using the *postMessage* technique.
It is exposing a very thin API that allows you to set the configuration and perform basic tasks (ie. clearing the session).
Instead, the Service will perform the real job by handling all HTTP requests.
The more powerful way to accomplish this goal is to implement it within a *serviceWorker* that allows to intercept the requests, furthermore one Service can handle multiple Client instances that are sharing the same origin (ie. when using your app in multiple tabs concurrently).

### Legacy mode

When the *ServiceWorker* is not available (ie. the browser lacks support for it) the *legacy package* will imitate the Service behavior. This is done by hooking the *XHR* and *fetch* components of the browser in order to intercept HTTP requests.
Even if this approach has some caveats (of course it's not ideal) it should work without issues with the majority of 3rd-party libraries because all standardized interfaces are strongly respected, also it has no issue when polyfills are needed.

### Packages

This library provides the following packages:

- [src/client.js](./src/client.js) is imported by `legacy` and `default`. You should not import this package directly, unless you need to make a custom bundle of *proauth.js* that may be usefull if you want build it within your app.
- [src/legacy.js](./src/legacy.js) is required when a browser has no *serviceWorker* support. It loads stuff to work in *legacy mode* and bootstraps the `client`.
- [src/default.js](./src/default.js) is required for the normal operational mode. It can register the `service-worker` and when ready bootstraps the `client`.
- [src/service-worker.js](./src/service-worker.js) is required for the normal operatinal mode and is used in the *serviceWorker* context. You can use it directly or can include it in your own *serviceWorker* distribution.

## Compatibility

### Default (ES6 + ServiceWorker)

* Chrome >= 45
* Firefox >= 51
* Android Chrome >= 6

### Legacy (ES5)

* Chrome >= 45
* Firefox >= 47
* Android Chrome >= 6
* Safari >= 10

## Testing

### Requirements

* **Node.js** >= 4
* **Yarn** >= 0.21

### First setup

* `yarn install`
* `cp saucelabs.config.local.json.dist saucelabs.config.local.json` and customize it

### Run

Simply execute `npm test`.

---

## Demo

To try the code, open the **demo.html** file in a browser and paste the following code in the console:

```
proauth.client.login("username", "password").then(r => console.log('Response', r), r => console.log('Error', r))

fetch('/user/me').then(function(response) {
  console.log(response.type); // basic
  console.log(response.json());
})
```

> Assuming that on the same host an OAuth2.0 authorization endpoint is available on `/oauth` and `/user/me` returns some data

---


CHANGELOG.md
<a name="0.3.2"></a>
## [0.3.2](https://github.com/fabbricadigitale/proauth.js/compare/v0.3.1...v0.3.2) (2017-07-17)


### Docs

* Browser matrix badge ([377f61508307730b9fcbb0e5aba66e2e2bc8401c](https://github.com/fabbricadigitale/proauth.js/commit/377f61508307730b9fcbb0e5aba66e2e2bc8401c))
* Update compatibility ([fdf3d3e5afd235d7481665a0caa76b4ba31cfc62](https://github.com/fabbricadigitale/proauth.js/commit/fdf3d3e5afd235d7481665a0caa76b4ba31cfc62))

### Fix

* Concurrency testing  ([1ad94c57948baf007f765c1d02baeb4a8a17e93e](https://github.com/fabbricadigitale/proauth.js/commit/1ad94c57948baf007f765c1d02baeb4a8a17e93e)), closes [#28](https://github.com/fabbricadigitale/proauth.js/issues/28)
* Ensure SauceLabs reporter config is not overridden ([316f79930937bc10a556c7d08df3c346f1350e97](https://github.com/fabbricadigitale/proauth.js/commit/316f79930937bc10a556c7d08df3c346f1350e97))
* Execute SauceLabs only on internal PRs ([804985785f08fba81ad8506b7177da3f513bd8be](https://github.com/fabbricadigitale/proauth.js/commit/804985785f08fba81ad8506b7177da3f513bd8be))
* Race condition in tests  ([184cdc20d51db2523431499a7aca8892fa08aaa9](https://github.com/fabbricadigitale/proauth.js/commit/184cdc20d51db2523431499a7aca8892fa08aaa9)), closes [#28](https://github.com/fabbricadigitale/proauth.js/issues/28)
* Tests for Safari 10 ([342b1e589da9939bbaa0d4b024d6e95d3553c42d](https://github.com/fabbricadigitale/proauth.js/commit/342b1e589da9939bbaa0d4b024d6e95d3553c42d))

### New

* Run tests on Saucelabs ([12ad9d818bc96799582692715d5171d8e633c3f5](https://github.com/fabbricadigitale/proauth.js/commit/12ad9d818bc96799582692715d5171d8e633c3f5))

### Update

* Add Android tests ([d85f889bac6a766ea0d66c31506ac4069346e64f](https://github.com/fabbricadigitale/proauth.js/commit/d85f889bac6a766ea0d66c31506ac4069346e64f))
* Improve SauceLabs fault tolerance  ([16b4087e722e69df5d2e66b726b5caa1b0162348](https://github.com/fabbricadigitale/proauth.js/commit/16b4087e722e69df5d2e66b726b5caa1b0162348))
* Refactor Saucelabs config to allow local testing ([418db72a6e0d8cd837fa8adabfad77a1c738446d](https://github.com/fabbricadigitale/proauth.js/commit/418db72a6e0d8cd837fa8adabfad77a1c738446d))
* Refactor SauceLabs configuration ([6589707d0ec6be9559dffa246076d0100ca965e2](https://github.com/fabbricadigitale/proauth.js/commit/6589707d0ec6be9559dffa246076d0100ca965e2))



<a name="0.3.1"></a>
## [0.3.1](https://github.com/fabbricadigitale/proauth.js/compare/v0.3.0...v0.3.1) (2017-06-22)


### Build

* Avoid race condition when making lib dir ([2ab7aae3dfb7698c42682fbcc503685e318fe59a](https://github.com/fabbricadigitale/proauth.js/commit/2ab7aae3dfb7698c42682fbcc503685e318fe59a))

### Fix

* Override mimetype also when no "content-type" header was received ([6ae208434d9747e11a9ae2e7fcfeeda945971458](https://github.com/fabbricadigitale/proauth.js/commit/6ae208434d9747e11a9ae2e7fcfeeda945971458))

### New

* Client whenReady promise ([a0a1567067043dc25aef41afa1562ce03c7f4cd7](https://github.com/fabbricadigitale/proauth.js/commit/a0a1567067043dc25aef41afa1562ce03c7f4cd7))
* Use native Karma webserver to avoid golang dependency ([3cc532ee676bb611494946247b72f7183297fb5a](https://github.com/fabbricadigitale/proauth.js/commit/3cc532ee676bb611494946247b72f7183297fb5a))

### Update

* Add linting to tests ([f622aa021d295fc2d930ce5f32c22384a9698da0](https://github.com/fabbricadigitale/proauth.js/commit/f622aa021d295fc2d930ce5f32c22384a9698da0))
* Keywords in package.json ([9c41de0acdcf2fc18d9fcf968e8355ebf459731b](https://github.com/fabbricadigitale/proauth.js/commit/9c41de0acdcf2fc18d9fcf968e8355ebf459731b))
* Lint correctly every file under "src" folder ([980f00c8e3eb68c7fcb76ca56cf10a6930c95d8b](https://github.com/fabbricadigitale/proauth.js/commit/980f00c8e3eb68c7fcb76ca56cf10a6930c95d8b))
* Refactor config ([dba7d72c9360dc2f886116ffca5d76c0db8ebb73](https://github.com/fabbricadigitale/proauth.js/commit/dba7d72c9360dc2f886116ffca5d76c0db8ebb73))
* Tests refactoring  ([0de381b334c5d0711f608c8e619d12b593323ef8](https://github.com/fabbricadigitale/proauth.js/commit/0de381b334c5d0711f608c8e619d12b593323ef8)), closes [#22](https://github.com/fabbricadigitale/proauth.js/issues/22)



<a name="0.3.0"></a>
# 0.3.0 (2017-06-15)


### Breaking

* Rejection on Client.login() error ([4ce9b8e37fc7d9262b8395276411f65bea3b1513](https://github.com/fabbricadigitale/proauth.js/commit/4ce9b8e37fc7d9262b8395276411f65bea3b1513))

### Build

* Autoinstall deps ([6cbbf02f24420ccb6527c8478a9d4f4b66fd341e](https://github.com/fabbricadigitale/proauth.js/commit/6cbbf02f24420ccb6527c8478a9d4f4b66fd341e))
* Build script ([497179a128ec327320acd996d59fc19f203a48d3](https://github.com/fabbricadigitale/proauth.js/commit/497179a128ec327320acd996d59fc19f203a48d3))
* Build system refinements ([c71998d08393968d7130bb689d450b753e84b5f9](https://github.com/fabbricadigitale/proauth.js/commit/c71998d08393968d7130bb689d450b753e84b5f9))
* Cross-platform CHANGELOG editor ([9a8a69ff6082c2f86b27299df52e80e7b87195d4](https://github.com/fabbricadigitale/proauth.js/commit/9a8a69ff6082c2f86b27299df52e80e7b87195d4))
* External source maps for production ([52217c8a8f1be348899c4de5198cc8cd963a5b37](https://github.com/fabbricadigitale/proauth.js/commit/52217c8a8f1be348899c4de5198cc8cd963a5b37))
* Fresh build system ([5b1ac7ef47ed2029ef073ccefbd972bebccb357d](https://github.com/fabbricadigitale/proauth.js/commit/5b1ac7ef47ed2029ef073ccefbd972bebccb357d))
* Include sources within distribution ([b2143f684a5d8baae4ddc8b918015b18bf459e6f](https://github.com/fabbricadigitale/proauth.js/commit/b2143f684a5d8baae4ddc8b918015b18bf459e6f))
* New simplified build system ([0b5d5a6c1b3de2740fe5f41440afa8ba6088cfea](https://github.com/fabbricadigitale/proauth.js/commit/0b5d5a6c1b3de2740fe5f41440afa8ba6088cfea))
* Postinstall script again ([82b82ebb370343e5864b8d0111cfe60e47a3810d](https://github.com/fabbricadigitale/proauth.js/commit/82b82ebb370343e5864b8d0111cfe60e47a3810d))

karma-options.js
"use strict"

const oauthFactory = function (config) {
  return function (request, response, next) {
    if (request.url === "/oauth") {
      let body = "";
      request.on("data", function (chunk) {
        body += chunk.toString("utf8")
      })
      request.on("end", function () {
        if (body) {
          const requestBodyJson = JSON.parse(body)

          if (requestBodyJson.username === "user" &&
            requestBodyJson.password === "qwerty" &&
            requestBodyJson.grant_type === "password" &&
            requestBodyJson.client_id === "test") {
            return response.end(`{"access_token":"tkn0.1234567890","expires_in":3600,"token_type":"bearer","scope":null,"refresh_token":"rfrsh0.0987654321"}`)
          } else if (requestBodyJson.refresh_token === "rfrsh0.0987654321" &&
            requestBodyJson.grant_type === "refresh_token" &&
            requestBodyJson.client_id === "test") {
            return response.end(`{"access_token":"tkn1.1234567890","expires_in":3600,"token_type":"bearer","scope":null,"refresh_token":"rfrsh1.0987654321"}`)
          }
        }

        response.writeHead(401)
        return response.end(`{"title": "invalid_grant","status":401}`)
      })

    } else {
      return next()
    }
  }
}

const returnAuthorizationHeader = function (config) {
  return function (request, response, next) {
    if (request.url === "/return-authorization-header") {
      return response.end(request.headers.authorization)
    } else {
      return next()
    }
  }
}

const returnEmptyResponse = function (config) {
  return function (request, response, next) {
    if (request.url === "/return-empty-response") {
      return response.end("")
    } else {
      return next()
    }
  }
}

const returnError401 = function (config) {
  return function (request, response, next) {
    if (request.url === "/return-error-401") {
      response.writeHead(401)
      return response.end(`{"title": "invalid_token","status":401}`)
    } else {
      return next()
    }
  }
}


const returnError401Concurrently = function (config) {
  const pendingUa = {}
  const requestMap = {}
  const concurrencyLimit = 3
  return function (request, response, next) {
    if (request.url.startsWith("/return-error-401-with-concurrency-3")) {
      const ua = request.headers["user-agent"]
      let count = 0

      // Only increase pending UA if this is not the 2nd request (i.e. the request with token re-negotiated by proauth)
      if (!requestMap[ua + request.url]) {
        count = pendingUa[ua] = pendingUa[ua] ? (pendingUa[ua] + 1) : 1
      }

      requestMap[ua + request.url] = true

      const intId = setInterval(function () {
        if (count >= concurrencyLimit) {
          count = pendingUa[ua] = 0
          clearInterval(intId)

          response.writeHead(401)
          return response.end(`{"title": "invalid_token","status":401}`)
        } else {
          count = pendingUa[ua]
        }
      }, 500)
    } else {
      return next()
    }
  }
}

const returnInvalidXML = function (config) {
  return function (request, response, next) {
    if (request.url === "/return-invalid-xml") {
      response.setHeader("Content-Type", "text/xml")
      return response.end("invalid_xml")
    } else {
      return next()
    }
  }
}

const returnSomeHeadersFactory = function (config) {
  return function (request, response, next) {
    if (/^\/return-some-headers[?]?\w*/.test(request.url)) {

      // Header purposely not in alphabetical order and upper/lowercased
      response.setHeader("Key-A", "Value-A")
      response.setHeader("Key-B", "Value-B")
      response.setHeader("Key-c", "Value-c")
      response.setHeader("key-e", "value-e")
      response.setHeader("KEY-D", "VALUE-D")

      return response.end("")
    } else {
      return next()
    }
  }
}

const simulateTokenExpiredFactory = function (config) {
  return function (request, response, next) {
    if (request.url === "/simulate-token-expired") {

      if (request.headers.authorization === "bearer tkn0.1234567890") {
        response.writeHead(401)

        return response.end(`{"title": "token_expired","status":401}`)
      } else {
        return response.end(request.headers.authorization)
      }
    } else {
      return next()
    }
  }
}

const sleepFactory = function (config) {
  return function (request, response, next) {
    if (request.url === "/sleep") {
      setTimeout(function () {
        return response.end("")
      }, 1000)
    } else {
      return next()
    }
  }
}

const customLaunchers = {
  sl_chrome_latest_winXP: {
    base: "SauceLabs",
    browserName: "chrome",
    platform: "Windows XP",
    version: "latest"
  },
  sl_chrome_latest_win7: {
    base: "SauceLabs",
    browserName: "chrome",
    platform: "Windows 7",
    version: "latest"
  },
  sl_chrome_latest_win8: {
    base: "SauceLabs",
    browserName: "chrome",
    platform: "Windows 8",
    version: "latest"
  },
  sl_chrome_latest_win10: {
    base: "SauceLabs",
    browserName: "chrome",
    platform: "Windows 10",
    version: "latest"
  },

  sl_chrome_latest_linux: {
    base: "SauceLabs",
    browserName: "chrome",
    platform: "Linux",
    version: "latest"
  },

  sl_chrome_latest_OSX10_8: {
    base: "SauceLabs",
    browserName: "chrome",
    platform: "OS X 10.8",
    version: "latest"
  },
  sl_chrome_latest_macOS10_12: {
    base: "SauceLabs",
    browserName: "chrome",
    platform: "MacOS 10.12",
    version: "latest"
  },

  sl_firefox_47_win7: {
    base: "SauceLabs",
    browserName: "firefox",
    platform: "Windows 7",
    version: "47"
  },
  sl_firefox_52_win10: {
    base: "SauceLabs",
    browserName: "firefox",
    platform: "Windows 10",
    version: "52"
  },
  sl_firefox_latest_winXP: {
    base: "SauceLabs",
    browserName: "firefox",
    platform: "Windows XP",
    version: "latest"
  },
  sl_firefox_latest_win7: {
    base: "SauceLabs",
    browserName: "firefox",
    platform: "Windows 7",
    version: "latest"
  },
  sl_firefox_latest_win8: {
    base: "SauceLabs",
    browserName: "firefox",
    platform: "Windows 8",
    version: "latest"
  },
  sl_firefox_latest_win10: {
    base: "SauceLabs",
    browserName: "firefox",
    platform: "Windows 10",
    version: "latest"
  },

  sl_firefox_latest_linux: {
    base: "SauceLabs",
    browserName: "firefox",
    platform: "Linux",
    version: "latest"
  },

  sl_firefox_latest_OSX10_8: {
    base: "SauceLabs",
    browserName: "firefox",
    platform: "OS X 10.8",
    version: "latest"
  },

  sl_firefox_latest_macOS10_12: {
    base: "SauceLabs",
    browserName: "firefox",
    platform: "MacOS 10.12",
    version: "latest"
  },

  sl_edge_latest_win10: {
    base: "SauceLabs",
    browserName: "microsoftedge",
    platform: "Windows 10",
    version: "latest"
  },

  sl_edge_14_win10: {
    base: "SauceLabs",
    browserName: "microsoftedge",
    platform: "Windows 10",
    version: "14"
  },

  sl_edge_15_win10: {
    base: "SauceLabs",
    browserName: "microsoftedge",
    platform: "Windows 10",
    version: "15"
  },

  sl_ie_11_win7: {
    base: "SauceLabs",
    browserName: "internetexplorer",
    platform: "Windows 7",
    version: "11"
  },

  sl_safari_latest_macos10_12: {
    base: "SauceLabs",
    browserName: "safari",
    platform: "MacOS 10.12",
    version: "latest"
  },

  sl_safari_10_macos10_12: {
    base: "SauceLabs",
    browserName: "safari",
    platform: "MacOS 10.12",
    version: "10"
  },

  sl_safari_9_OSX10_11: {
    base: "SauceLabs",
    browserName: "safari",
    platform: "OS X 10.11",
    version: "9"
  },

  sl_chrome_latest_android6_0: {
    base: "SauceLabs",
    browserName: "chrome",
    deviceName: "Android Emulator",
    platformName: "Android",
    platformVersion: "6.0"
  },
  sl_chrome_latest_android7_0: {
    base: "SauceLabs",
    browserName: "chrome",
    deviceName: "Android GoogleAPI Emulator",
    platformName: "Android",
    platformVersion: "7.0"
  },
  sl_browser_4_4_android4_4: {
    base: "SauceLabs",
    browserName: "Browser",
    deviceName: "Android Emulator",
    platformName: "Android",
    platformVersion: "4.4",
    appiumVersion: "1.6.5"
  },

  sl_safari_10_2_ios_10_2: {
    base: "SauceLabs",
    browserName: "Safari",
    deviceName: "iPhone Simulator",
    platformName: "iOS",
    platformVersion: "10.2",
    appiumVersion: "1.6.5"
  },
  sl_safari_10_3_ios_10_3: {
    base: "SauceLabs",
    browserName: "Safari",
    deviceName: "iPhone Simulator",
    platformName: "iOS",
    platformVersion: "10.3",
    appiumVersion: "1.6.5"
  }

}

const options = {

    // base path that will be used to resolve all patterns (eg. files, exclude)
    basePath: "",

    // frameworks to use
    // available frameworks: https://npmjs.org/browse/keyword/karma-adapter
    frameworks: ["jasmine"],

    // list of files / patterns to load in the browser
    files: [
    ],

    // list of files to exclude
    exclude: [
    ],

    // preprocess matching files before serving them to the browser
    // available preprocessors: https://npmjs.org/browse/keyword/karma-preprocessor
    preprocessors: {
    },

    // test results reporter to use
    // possible values: "dots", "progress"
    // available reporters: https://npmjs.org/browse/keyword/karma-reporter
    reporters: [
      "spec",
      "progress",
      "coverage"
    ],
    // web server port
    port: 9876,

    // enable / disable colors in the output (reporters and logs)
    colors: true,

    // enable / disable watching file and executing tests whenever any file changes
    autoWatch: true,

    // start these browsers
    // available browser launchers: https://npmjs.org/browse/keyword/karma-launcher
    browsers: ["Chrome", "Firefox"],

    proxies: {
      "/service-worker.js": "/base/lib/service-worker.es2015.js"
    },

    plugins: [
      "karma-*",
      { "middleware:oauth": ["factory", oauthFactory] },
      { "middleware:return-authorization-header": ["factory", returnAuthorizationHeader] },
      { "middleware:return-empty-response": ["factory", returnEmptyResponse] },
      { "middleware:return-error-401": ["factory", returnError401] },
      { "middleware:return-error-401-with-concurrency-3": ["factory", returnError401Concurrently] },
      { "middleware:return-invalid-xml": ["factory", returnInvalidXML] },
      { "middleware:return-some-headers": ["factory", returnSomeHeadersFactory] },
      { "middleware:simulate-token-expired": ["factory", simulateTokenExpiredFactory] },
      { "middleware:sleep": ["factory", sleepFactory] }
    ],

    middleware: [
      "oauth",
      "return-authorization-header",
      "return-empty-response",
      "return-error-401",
      "return-error-401-with-concurrency-3",
      "return-invalid-xml",
      "return-some-headers",
      "simulate-token-expired",
      "sleep"
    ],

    browserNoActivityTimeout: 600000,

    retryLimit: 5,

    customLaunchers,

    // Continuous Integration mode
    // if true, Karma captures browsers, runs the tests and exits
    singleRun: true,

    // Concurrency level
    // how many browser should be started simultaneous
    concurrency: 1

  }

if (process.env.TEST_SAUCELABS) {
  options.reporters.push("saucelabs")
  options.sauceLabs = require("./saucelabs.config.js")
}

module.exports = options
